google-cloud-storage/src/test/java/com/google/cloud/storage/spi/v1/HttpStorageRpcTest.java

```diff
@@ -0,0 +1,178 @@
+import com.google.cloud.http.HttpTransportOptions;
+import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.StorageOptions;
+import com.google.cloud.storage.spi.StorageRpcFactory;
+import com.google.common.io.ByteStreams;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+import java.util.zip.GZIPInputStream;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+public class HttpStorageRpcTest {
+
+  // object under test, created before each test case
+  private HttpStorageRpc rpc;
+
+  // Objects for a test case: one object per one RPC mock.
+  // A test creates as many MockData objects as it issues RPC calls.
+  private Queue<MockData> testMockData;
+
+  private static final String BASE_URL = "https://storage.googleapis.com/storage/v1/b";
+  private static final String URL_PROJECT = "project=projectId&projection=full";
+
+  private static final Answer<Object> UNEXPECTED_CALL_ANSWER =
+      new Answer<Object>() {
+        @Override
+        public Object answer(InvocationOnMock invocation) {
+          throw new IllegalArgumentException(
+              "Unexpected call of "
+                  + invocation.getMethod()
+                  + " with "
+                  + Arrays.toString(invocation.getArguments()));
+        }
+      };
+
+  private static final ApiClock TIME_SOURCE =
+      new ApiClock() {
+        @Override
+        public long nanoTime() {
+          return 42_000_000_000L;
+        }
+
+        @Override
+        public long millisTime() {
+          return 42_000L;
+        }
+      };
+
+  private static final StorageRpcFactory RPC_FACTORY_MOCK =
+      mock(StorageRpcFactory.class, UNEXPECTED_CALL_ANSWER);
+
+  /**
+   * Adds mock parameters for the RPC call.
+   *
+   * @param method expected HTTP method
+   * @param url expected URL
+   * @param response HTTP response object to be returned
+   * @return a holder object which will contain the Http Request made by RPC
+   */
+  private RpcRequestHolder mockResponse(String method, String url, LowLevelHttpResponse response) {
+    RpcRequestHolder holder = new RpcRequestHolder();
+    testMockData.add(new MockData(method, url, response, holder));
+    return holder;
+  }
+
+  private static String getUnzippedContent(LowLevelHttpRequest request) {
+    try {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      request.getStreamingContent().writeTo(outputStream);
+      GZIPInputStream gzipInputStream =
+          new GZIPInputStream(new ByteArrayInputStream(outputStream.toByteArray()));
+      return new String(ByteStreams.toByteArray(gzipInputStream), UTF_8);
+    } catch (IOException e) {
+      throw new Error(e);
```

Comment:
RuntimeException, or better yet just bubble the IOException

Reply:
I selected Error because `An Error is a subclass of Throwable that indicates serious problems that a reasonable application should not try to catch. Most such errors are abnormal conditions.`
IOException thrown from this try-catch block will signal about problems in the core Java libraries, not in the Google libraries under test.

Reply:
I replaced `Error` with `AssertionError` which `JUnit.fail()` throws. I can't replace with `fail()` because the method returns a value.